reportsrv/reportsrv.go

```diff
@@ -0,0 +1,194 @@
+	r.Get("/report/{analysis_id}/*", s.serveReport)
+}
+
+func (s *Server) serveReport(w http.ResponseWriter, r *http.Request) {
+	aID := pacta.AnalysisID(chi.URLParam(r, "analysis_id"))
+	if aID == "" {
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		return
+	}
+	ctx := r.Context()
+
+	artifacts, err := s.db.AnalysisArtifactsForAnalysis(s.db.NoTxn(ctx), aID)
+	if err != nil {
+		s.logger.Error("failed to load artifacts for analysis", zap.String("analysis_id", string(aID)), zap.Error(err))
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+	var blobIDs []pacta.BlobID
+	for _, a := range artifacts {
+		blobIDs = append(blobIDs, a.Blob.ID)
+	}
+
+	blobs, err := s.db.Blobs(s.db.NoTxn(ctx), blobIDs)
+	if err != nil {
+		s.logger.Error("failed to load blobs", zap.Strings("blob_ids", asStrs(blobIDs)), zap.Error(err))
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	subPath := strings.TrimPrefix(r.URL.Path, "/report/"+string(aID))
+	if strings.HasPrefix(subPath, "/") {
+		subPath = subPath[1:]
+	}
+	if subPath == "" {
+		subPath = "index.html"
+	}
+
+	for _, a := range artifacts {
+		// Container is just 'reports', we can ignore that.
+		b, ok := blobs[a.Blob.ID]
+		if !ok {
+			s.logger.Error("no blob loaded for blob ID", zap.String("analysis_artifact_id", string(a.ID)), zap.String("blob_id", string(a.Blob.ID)))
+			continue
+		}
+		uri := string(b.BlobURI)
+		_, path, ok := blob.SplitURI(s.blob.Scheme(), uri)
+		if !ok {
+			s.logger.Error("blob had invalid URI", zap.String("analysis_artifact_id", string(a.ID)), zap.String("blob_uri", uri))
+			continue
+		}
+		_, uriPath, ok := strings.Cut(path, "/")
+		if !ok {
+			s.logger.Error("path had no UUID prefix", zap.String("analysis_artifact_id", string(a.ID)), zap.String("blob_uri", uri), zap.String("blob_path", path))
+			continue
+		}
+
+		// This isn't the asset we're looking for.
+		if uriPath != subPath {
+			continue
+		}
+
+		// Load the blob from blob storage and copy it byte-for-byte to the HTTP response.
+		r, err := s.blob.ReadBlob(ctx, uri)
+		if err != nil {
+			http.Error(w, "failed to read blob: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+		defer r.Close()
+
+		w.Header().Set("Content-Type", fileTypeToMIME(b.FileType))
+		if _, err := io.Copy(w, r); err != nil {
+			s.logger.Error("failed to read/write blob", zap.String("blob_uri", uri), zap.Error(err))
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+		return
+	}
+
+	s.logger.Error("unknown report asset requested for analysis", zap.String("analysis_id", string(aID)), zap.String("req_path", r.URL.Path), zap.String("asset_path", subPath))
+	http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+	return
+}
+
+func fileTypeToMIME(ft pacta.FileType) string {
```

Comment:
I guess this is one good reason to have FT saved!

Reply:
Indeed! As discussed, I'll update the parent PR to store unknown as `'unknown'`